test/functional/test_framework/test_framework.py

```diff
@@ -181,33 +196,25 @@ def main(self):
         self.network_thread = NetworkThread()
         self.network_thread.start()
 
-        success = TestStatus.FAILED
+        self.success = TestStatus.FAILED
 
-        try:
-            if self.options.usecli:
-                if not self.supports_cli:
-                    raise SkipTest("--usecli specified but test does not support using CLI")
-                self.skip_if_no_cli()
-            self.skip_test_if_missing_module()
-            self.setup_chain()
-            self.setup_network()
-            self.run_test()
-            success = TestStatus.PASSED
-        except JSONRPCException:
-            self.log.exception("JSONRPC error")
-        except SkipTest as e:
-            self.log.warning("Test Skipped: %s" % e.message)
-            success = TestStatus.SKIPPED
-        except AssertionError:
-            self.log.exception("Assertion failed")
-        except KeyError:
-            self.log.exception("Key error")
-        except Exception:
-            self.log.exception("Unexpected exception caught during testing")
-        except KeyboardInterrupt:
-            self.log.warning("Exiting after keyboard interrupt")
+        if self.options.usecli:
+            if not self.supports_cli:
+                raise SkipTest("--usecli specified but test does not support using CLI")
+            self.skip_if_no_cli()
+        self.skip_test_if_missing_module()
+        self.setup_chain()
+        self.setup_network()
 
-        if success == TestStatus.FAILED and self.options.pdbonfailure:
+    def shutdown(self, e=None, exit=False):
+        """Call this method to shutdown the test object and optionally handle an exception."""
+
+        if e != None:
```

Comment:
I think it makes more logical sense to have the exception handling code in the `main()` function, and then just start `shutdown()` at the `if self.success == TestStatus.FAILED and ...` line

Reply:
I am sure I fully understand your suggestion. If we move error handling into main, this is code that needs to be replicated in TestWrapper.

Reply:
why not:

```py
        try:
            self.setup()
            self.run_test()
        except BaseException as e:
            handle_exception(e)
        finally:
            exit_code = self.shutdown()
            sys.exit(exit_code)
```